flows/utils.py

```diff
@@ -526,3 +568,61 @@ async def map_as_sub_flow(
                 failures.append(result)
 
     return successes, failures
+
+
+def default_desc(tasks, results) -> str:
+    return f"Finished task {len(results)} of {len(tasks)}"
+
+
+@overload
+async def gather_and_report(
+    tasks: Sequence[Awaitable[T]],
+    return_exceptions: Literal[True],
+    key: str,
+    desc_create: str,
+    desc_update_fn: Callable[[Sequence[Any], list[Any]], str] = default_desc,
+) -> Sequence[T | Exception]: ...
+
+
+@overload
+async def gather_and_report(
+    tasks: Sequence[Awaitable[T]],
+    return_exceptions: Literal[False],
+    key: str,
+    desc_create: str,
+    desc_update_fn: Callable[[Sequence[Any], list[Any]], str] = default_desc,
+) -> Sequence[T]: ...
+
+
+async def gather_and_report(
+    tasks: Sequence[Awaitable[T]],
+    return_exceptions: bool,
+    key: str,
+    desc_create: str,
+    desc_update_fn: Callable[[Sequence[Any], list[Any]], str] = default_desc,
+) -> Sequence[T] | Sequence[T | Exception]:
+    progress_artifact_id: UUID = await create_progress_artifact(
```

Comment:
I get some linting issues on this line and 622: 

> "UUID" is not awaitable
>   "UUID" is incompatible with protocol "Awaitable[_T_co@Awaitable]"
>     "__await__" is not presentbasedpyright[reportGeneralTypeIssues](https://docs.basedpyright.com/v1.29.4/configuration/config-files/#reportGeneralTypeIssues)

Reply:
It's the same as for the other `create_foo_artifact` calls, annoyingly.

Reply:
I've suggested, in Slack, walking through the code at some point.